src/Orleans.Serialization/Buffers/Adaptors/PooledArrayBufferWriter.cs

```diff
@@ -201,7 +201,12 @@ private SequenceSegment(int length)
         {
             if (length <= MinimumBlockSize)
             {
+#if NET6_0_OR_GREATER
                 var pinnedArray = GC.AllocateUninitializedArray<byte>(MinimumBlockSize, pinned: true);
+#else
+                var pinnedArray = new byte[MinimumBlockSize];
+                GCHandle.Alloc(pinnedArray, GCHandleType.Pinned);
```

Comment:
This is probably the biggest functional difference (other changes our simple signature adjustments or conditionally rolling back to an earlier version of the code). Not holding onto that handle feels dangerous, but it looked like the array would simply be recycled in this case.

Reply:
Certainly could; however, I'm not seeing a place where we would actually free it. It looks look returning a segment just returns in to the `Blocks` bag and retains the memory in the `IsStandardSize` case (only case that explicitly pins memory). If there a limit on how many returned segments were kept around then freeing would definitely be important.

Reply:
This can potentially leak these `GCHandle`s, can't it? I think we should just not pin these. The arrays should never be accessed in an unsafe manner anyway. Putting them on the PinnedObjectHeap is mostly a way to reduce potential fragmentation for these large (but not LOH large), pooled arrays.

Reply:
Yeah, it could leak.
Not pinning is fine. Would you like me to update the net6+ case to also not pin? As is, I believe they have the same leaking potential.

Reply:
The .NET 6.0 case wont leak: the GC can claim objects in the POH, but it won't move them.
The pinned `GCHandle` will leak, though, as far as I know. See the GCHandle doc comments:
```csharp
/// <summary>
/// Represents an opaque, GC handle to a managed object. A GC handle is used when an
/// object reference must be reachable from unmanaged memory.
/// </summary>
/// <remarks>
/// There are 4 kinds of roots:
/// Normal: Keeps the object from being collected.
/// Weak: Allows object to be collected and handle contents will be zeroed.
/// Weak references are zeroed before the finalizer runs, so if the
/// object is resurrected in the finalizer the weak reference is still zeroed.
/// WeakTrackResurrection: Same as Weak, but stays until after object is really gone.
/// Pinned - same as Normal, but allows the address of the actual object to be taken.
/// </remarks>
```